schema-sandbox/schema.json

```
@@ -64,6 +79,11 @@
                     "type": "string"
                 }
             }
+        },
+        "url": {
+            "type": "string",
+            "format": "uri",
+            "pattern": "^(?:(?:https?|ftp)://)(?:\\S+(?::\\S*)?@)?(?:(?!(?:10|127)(?:\\.\\d{1,3}){3})(?!(?:169\\.254|192\\.168)(?:\\.\\d{1,3}){2})(?!172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})(?:[1-9]\\d?|1\\d\\d|2[01]\\d|22[0-3])(?:\\.(?:1?\\d{1,2}|2[0-4]\\d|25[0-5])){2}(?:\\.(?:[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-4]))|(?:(?:[a-z\\u00a1-\\uffff0-9]-?)*[a-z\\u00a1-\\uffff0-9]+)(?:\\.(?:[a-z\\u00a1-\\uffff0-9]-?)*[a-z\\u00a1-\\uffff0-9]+)*(?:\\.(?:[a-z\\u00a1-\\uffff]{2,})))(?::\\d{2,5})?(?:/\\S*)?$"
```

I'd propose the following:

Remove the regex here, and for 1.2.0, propose lax validation with regexes only where they are simple and make sense. In my opinion a lax schema shouldn't be the place to validate actual user values in cases where there is no proven correct regex, e.g., for URLS (see e.g., [demoing different patterns in search of the perfect URL regex](https://mathiasbynens.be/demo/url-regex)). Also, relaxing the schema doesn't "break API" as more strictly validated values from older versions will still be valid.

Then, do a *second* schema for strict validation for a `1.3.0` (or other feature) version in the future, where we can introduce stricter validation of user values with best effort regexes. This could potentially "break API" (when we switch to a better regex", so could be a 2.0.0 version.

Also, let's say, "downstream services" will most likely be happy with the lax schema for reasons of performance ;).

The laxity of the schema should be documented somewhere, but doesn't have to be very prominent.

What do you think @jspaaks?

```suggestion
            "format": "uri"
```

---

Ah, also `"format": "uri"` goes some way to ensure validity, doesn't it?

---

Was thinking whether `uri-reference` may be the better format, but we don't want to accept relative references here, do we?

---

In terms of simplification, I was thinking we could replace the current, complicated regex with a simpler one that will work basically the same: 

```
^(https|http|ftp|sftp)://
```

Combined with the `format: uri`, this ensures you get the subset of uris that start with `http://`, `https://`, `ftp://` or `sftp://`. The old regex is difficult to read but it seems to support `http`, `https`, and `ftp` (but not `sftp`)

I made an issue about this, so we can just merge the current PR and then simplify later

---

Hm, so using the `uri` with the simplified regex format doesn't actually catch validation issues for malformed URIs such as `https://\uD800`. Probably due to a lack of support in the jsonschema package.

---

I think I can live with that though

---

Okay then, let's use the simple prefix regex.